### PR TITLE
Dialog component

### LINKED ISSUE #197: Dialog widget
**Enhancement**

Add a dialog widget

Related to #85

### PR DESCRIPTION
**Pull request by bitpshr**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

This PR introduces a Dialog component. The API is fairly minimal. Properties to control alignment, positioning, and sizing of the dialog were omitted in favor of embracing CSS. I'm expecting to add to the API as necessary features are identified through use. This PR also adds an `examples` folder with a per-widget example structure. Note that because component CSS hasn't been solidified, base structural styles have been temporarily included on the example dialog page itself.

Initial API:

```ts
closeable?: boolean;
enterAnimation?: string;
exitAnimation?: string;
modal?: boolean;
open?: boolean;
title?: string;
underlay?: boolean;
onOpen?(): void;
onRequestClose?(): void;
```

Resolves #197

### CODE REVIEW

File: src/components/dialog/createDialog.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	underlay?: boolean;
+};
+
+export interface DialogProperties extends WidgetProperties {
+	modal?: boolean;
+	open?: boolean;
+	title?: string;
+	underlay?: boolean;
+	onOpen?(): void;
+	onRequestClose?(): void;
```
Comment (agubler) [inline src/components/dialog/createDialog.ts]: do you think we should make this mandatory?

Comment (bitpshr) [inline src/components/dialog/createDialog.ts]: I was thinking about the use case (probably rare) where a dialog shouldn't be able to be closed by a user clicking the close button. Maybe an application wants to show a dialog during a network operation and hide it when it's complete. Now that I think about it, we probably want to leave `onRequestClose` as optional, but add a `closeable` boolean option that hides the close button.

Comment (bitpshr) [inline src/components/dialog/createDialog.ts]: Added.

File: src/components/dialog/createDialog.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+						onclick: this.onUnderlayClick
+					}),
+					v('div.main', {
+						enterAnimation: enterAnimation,
+						exitAnimation: exitAnimation
+					}, [
+						v('div.title', [
+							this.state.title ? this.state.title : null,
+							closeable ? v('div.close', {
+								innerHTML: '✕',
```
Comment (rishson) [inline src/components/dialog/createDialog.ts]: &times; instead of 'x'?

File: src/components/dialog/createDialog.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			},
+
+			getChildrenNodes: function (this: Dialog): DNode[] {
+				const {
+					closeable = true,
+					enterAnimation = 'show',
+					exitAnimation = 'hide'
+				} = this.properties;
+
+				const children: DNode[] = [
```
Comment (rishson) [inline src/components/dialog/createDialog.ts]: Are we adding `a11y` concerns from the get-go?

**Comment by bitpshr** [comment]
Closing this for now, will clean up and submit a new PR once components and core are split (dojo/meta#134).

**Pull request closed without merging**